keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an application developer, I'm knowledgeable to know when my Pods are schedule-ready, but I know
+nothing about scheduling internals. It'd be ideal for me to only set/unset the "paused" flag,
+instead of developing scheduler plugins, to control the schedulability of my Pods.
+
+#### Story 2
+
+As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
+to guard the schedulability of my Pods.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+<<[UNRESOLVED]>>
+
+The feature will be guarded by a feature gate and disabled by default in alpha. When alpha, surely 
+the Pod API field needs to be disabled, but I'm open to hear feedback on whether to disable scheduler's
+ComponentConfig API as well as the extension point logic.
+
+If we disable it, then users cannot use the new extension point at all.
+If we don't disable it, people can still customize their enqueue behaviors, like wire it with a
+customized field (like an annotation key).
+
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED]>>
+
+As we don't write unschedulable condition to the Pod's spec if it carries `.status.schedulePaused=true`,
+it may provider better UX to add a "paused" column to the output of `kubectl get pod (-o wide)`.
+
+<<[/UNRESOLVED]>>
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
+should provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+A new API field `schedulePaused` will be added to Pod's status:
+
+```go
+type PodStatus struct {
+    SchedulePaused *bool // <-- New field.
```

Comment:
Will the field become non-mutable for a pod that has finished scheduling and has a nodeName assigned?

Reply:
In the context of a pod which can only be scheduled once, i would probably argue the expected behavior is the field is immutable after nodeName is set (prevent unexpected transitions early).

I had a question above about the behavior prior to scheduling that as similar

Reply:
I think validation should be even tighter on updates and only allow true -> false transitions, otherwise the risk of race conditions is much higher.

Reply:
Semantically, if a pod's `.spec.nodeName` is non-nil, it means the Pod has passed scheduling phase, so `schedulingPaused` is a no-op. So what's the net benefits of making it immutable after the pod's nodeName is set?

Given scheduled Pod can still get rejected by kubelet, esp. when multiple scheduler are in place, so in API Server's perspective, if we make the paused flag immutable, can we unset its nodeName AND flip the paused flag in one API call? (my guess is no)

Reply:
> Restricting the transition of the flag from true-to-false will help with that.

> In fact, another thing we should do is to reject nodeName update if the flag was set to true.

The case I described above was primarily the conflict among multi schedulers, so no matter w/ or w/o the true->false restriction, a scheduled Pod can somehow be rejected by kubelet. My point was, ops was able to "reprieve" this Pod by unsettling its nodeName (to nil) and flip schedulingPaused (to true), but with this restriction they cannot - deleting and recreating the Pod is the only way.

We may argue this is a hypothetical case, but better discuss here. If we decide not to support this, we should document it well.

Reply:
> The case I described above was primarily the conflict among multi schedulers, so no matter w/ or w/o the true->false restriction, a scheduled Pod can somehow be rejected by kubelet

I think we shouldn't use that case as a reference and a design pattern to follow. Multi schedulers is supported, but the recommended pattern should be for each scheduler to handle different workloads on different nodes (we need to make that clear in the docs if we don't already). So even for multi schedulers, Kubelet admission logic is acting as a last resort for protection, and ideally shouldn't be rejecting pods if the multiple schedulers where setup correctly.

> My point was, ops was able to "reprieve" this Pod by unsettling its nodeName (to nil) and flip schedulingPaused (to true), but with this restriction they cannot - deleting and recreating the Pod is the only way.

I would argue that nodeName should also be set once, but I don't think we can tighten validation now on this since it would be backward incompatible.

Reply:
A consensus we may reach is the one-time transition of _unscheduled_ Pod - you cannot flip the .schedulingPaused field via update back and forth, only from true->fase:

|                                     | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
|-------------------------------------|------------------------|------------------------|
| unscheduled Pod<br>(nil nodeName)   | ✅ create<br>❌ update   | ✅ create<br>✅ update   |

In terms of _scheduled_ Pod, I rethink a bit about the case mentioned above, to "revive" this Pod, we could "achieve" via:
- 1 API call: update **both**  nodeName (to nil) and schedulingPaused (to true), but this path is assumed to be blocked (❌ above) already.
- 2 API calls: firstly update schedulingPaused (to false), and then update nodeName (to nil). In this path, the first step is pretty misleading, as you'd reach a state that a scheduled Pod with schedulingPaused=true. So although the second call is allowed so far, the first API call looks problematic, and overall, we may abandon this path.

With the above being said, I think we should impose validation of two state changes:

- disallow a Pod's update on changing schedulingPaused to true.
- disallow any state that a Pod becomes schedulable while also carrying schedulingPaused=true.

Validation diagram is like this:

|                                     | schedulingPaused=true⏸️ | schedulingPaused=false▶️ |
|-------------------------------------|------------------------|-------------------------|
| unscheduled Pod<br>(nil nodeName)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
| scheduled Pod<br>(non-nil nodeName) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |

Reply:
I will update the design if ^^ makes sense to all of you.

Reply:
Re: blocking setting nodeName when scheduling is paused, I'm a bit concerned about blocking that because today a client is allowed to forcibly schedule a node, and that may not be a scheduler, but might be an admin.  Also, schedulers that are not aware of the scheduling paused field (have not been code updated) should behave predictably (i.e. we have to decide what the behavior is) when this field is added and a user uses it.

Scenario:

1. Cluster has two schedulers, default and custom
2. Some automation uses the custom scheduler to create pods
3. Cluster is upgraded to support `schedulingPaused`
4. Default scheduler updated to understand schedulingPaused, custom is not
5. A new workload of setting paused is added to the default scheduler (i.e. new consumer)

Can we make arguments for or against:

1. Allowing a human admin to force schedule a paused pod via setting nodeName without being aware of paused?
2. Preventing the custom scheduler from being broken by a human setting "paused" on one of the custom workload jobs (which puts the scheduler into a break loop)?

There is also the forward compatibility angle of adding this field for the "unaware" custom scheduler - we need to make an argument for or against support there from "does the addition of this field have unexpected outcomes for existing code that is unreasonable / dangerous"?

Reply:
> 1. Allowing a human admin to force schedule a paused pod via setting nodeName without being aware of paused?
> 2. Preventing the custom scheduler from being broken by a human setting "paused" on one of the custom workload jobs (which puts the scheduler into a break loop)?

I think we can have an in-tree admission controller to unset the "paused" field for scheduled Pod. cc/ @ahg-g 

The intention of "blocking setting nodeName when scheduling is paused" was to avoid discrepancy, which is "a scheduled pod, but carries a schedulingPaused state, which looks weird". If we can clear the "paused" field, we're good to always guarantee a consistent view.

Reply:
>  "a scheduled pod, but carries a schedulingPaused state, which looks weird"

Weird is vague :) - is this a better way to describe what you mean?

> Will clients be confused if it remains set?

How would the scheduler have to react in that case?  Should all scheduler impls have to track both fields?  Is a scheduler allowed to ignore "paused" (perhaps one that has not been upgraded yet)?  Forward compat was the reason that I was originally asking.

Will server side apply or other client apply loops be confused if the field remains set?  I.e. "paused" is the initial value, will a subsequent apply work as the user would expect?

Will a workload controller (statefulset, deployment) pod template transitioning from `schedulingPaused: true` to `schedulingPaused: false` result in redeployment of the pods underneath it?  If it doesn't, will the clearing of paused by the scheduler impact how the various workload controllers deploy pods?

> we're good to always guarantee a consistent view.

Consistency is good, but it's generally a default and can be overriden by ergonomics, forward compatibility, and careful consideration of the impact of adding the field.  We don't have a great set of rules for "how to think about adding fields to core objects like pods safely" and I'm trying to see if we can tease those out here in that absence.

Reply:
> Weird is vague :) - is this a better way to describe what you mean?

Semantically, a pod that has nodeName assigned indicates it has passed scheduling, so it doesn't make sense to have a pod in that state. 

We're going to add a column "scheduling paused" to `k get pod`. So if in the output of `k get pod`, it would show a pod has node name assigned, and also schedulingPaused=true, that would confuse the user a lot IMO.

> How would the scheduler have to react in that case? Should all scheduler impls have to track both fields? Is a scheduler allowed to ignore "paused" (perhaps one that has not been upgraded yet)? Forward compat was the reason that I was originally asking.

Default scheduler itself doesn't track the "paused" state if it's already scheduled. However, this doesn't mean we _should_ leave it this way. Integrators may rely on this field to compose their behavior. For example, when we built preemption feature, `.status.nomiantedNodeName` is set to indicate a pod is preempting other pods, and that field is supposed to be cleared when it gets scheduled. Well, in scheduler's POV, we _could_ claim we don't care about that state if it's scheduled (by default scheduler or by custom scheduler or enforced by admin), but Cluster Autoscaler does rely on that field to proceed their scale-up/down logic. This is the same case for "paused" state - make the state always accurate either by (1) disallowing update nodeName for "paused" pod or (2) actively clear the "paused" state for scheduled pod.

> Will server side apply or other client apply loops be confused if the field remains set?

Given we're gonna enforce a true->false on update, user would see the update as expected.

> Will a workload controller (statefulset, deployment) pod template transitioning from schedulingPaused: true to schedulingPaused: false result in redeployment of the pods underneath it? If it doesn't, will the clearing of paused by the scheduler impact how the various workload controllers deploy pods?

No, it doesn't yield redeployment. And it won't impact how the various workload controllers deploy pods either.

(BTW: scheduler doesn't clear the state, only controllers do. Or, "to-be-determined" admission controller in a particular circumstance.)

Reply:
> I think we can have an in-tree admission controller to unset the "paused" field for scheduled Pod. cc/ @ahg-g

I don't think we want to go that route, this field is not a status field to reconcile it in this fashion. 

What I was hoping to achieve with "blocking setting nodeName when scheduling is paused" is consistency on scheduling **behavior** across all schedulers and scenarios, meaning a pod will in all circumstances will not be assigned a node if it was paused. 

@smarterclayton consider the scenario where a feature is built using this flag, like statefulset migration, and we have a customer scheduler that wasn't updated to take this flag into account; then we have to select between two failure scenarios:

1) the statefulset migration feature being broken in an unexpected way if we don't block setting nodeName when pause=true
2) custom schedulers erroring out

I think the second failure is more tolerable because it doesn't put the cluster in an unexpected state.

Reply:
> I don't think we want to go that route

That was an idea based on supporting the scenario Clayton mentioned - a custom scheduler or admin force setting a nodeName while it's paused.

> I think the second failure is more tolerable because it doesn't put the cluster in an unexpected state.

+1.

Reply:
but I wouldn't push hard on it if others consider this backward incompatible.